yxi/axi-calyx/axi-combined-calyx.futil

```diff
@@ -0,0 +1,1239 @@
+        n_finished_last_trnsfr.in = (max_trnsfrs.out == curr_trnsfr_count.out) & wvalid.out & WREADY ? 1'b0;
+        n_finished_last_trnsfr.write_en = (max_trnsfrs.out == curr_trnsfr_count.out) & wvalid.out & WREADY ? 1'b1;
+
+
+
+        // we are done after handshake
+        bt_reg.in = wvalid.out & WREADY ? 1'b1;
+        bt_reg.in = !(wvalid.out & WREADY) ? 1'b0;
+        bt_reg.write_en = 1'b1;
+        do_write_transfer[done] = bt_reg.out;
+      }
+
+      group incr_curr_addr{
+          curr_addr_adder.left = 64'd1 ;
+          curr_addr_adder.right = curr_addr.out;
+          curr_addr.in = curr_addr_adder.out;
+          curr_addr.write_en = 1'b1;
+          incr_curr_addr[done] = curr_addr.done;
+      }
+
+      group incr_curr_trnsfr_count {
+          curr_trnsfr_count_adder.left = 8'd1;
+          curr_trnsfr_count_adder.right = curr_trnsfr_count.out;
+          curr_trnsfr_count.in = curr_trnsfr_count_adder.out;
+          curr_trnsfr_count.write_en = 1'b1;
+          incr_curr_trnsfr_count[done] = curr_trnsfr_count.done;
+      }
+
+  }
+  control{
+      seq{
+
+        reset_curr_addr;
+        //end writing to internal mem
+        init_n_finished_last_trnsfr;
+        while n_finished_last_trnsfr.out{
+          seq{
+              reset_bt;
+              do_write_transfer;
+              par{
+                incr_curr_addr;
+                incr_curr_trnsfr_count;
+              }
+          }
+        }
+      }
+  }
+}
+
+
+//We assume that all responses are OKAY because we dont have any error handling.
+//So basically this just sets BREADY high then lowers it
+component m_bresp_channel(
+  ARESET : 1,
+  BVALID : 1,
+  // We assume all writes are valid.
+  //BRESP : 2,
+) -> (
+  // NOTE: In general, according to ZipCPU we want xREADY signals to be registered
+  // because (IIRC) it helps avoid combinational loops between READY and VALID.
+  BREADY : 1,
+) {
+  cells{
+      bready = std_reg(1);
+      bt_reg = std_reg(1);
+
+  }
+  wires{
+      BREADY = bready.out;
+      group reset_bt_reg{
+          bt_reg.in = 1'b0;
+          bt_reg.write_en = 1'b1;
+          reset_bt_reg[done] = bt_reg.done;
+      }
+
+      // TODO(nathanielnrn): This is probably very unoptimal and takes multiple
+      // cycles to simply do a handshake. Can probably be much better
+      group block_transfer{
+          bready.in = !(BVALID & bready.out) ? 1'b1;
+          bready.in = BVALID & bready.out ? 1'b0;
+          bready.write_en = 1'b1;
+
+          bt_reg.in = bready.out & BVALID ? 1'b1;
+          bt_reg.in = !(bready.out & BVALID) ? 1'b0;
+          bt_reg.write_en = 1'b1;
+          block_transfer[done] = bt_reg.out;
+      }
+
+  }
+  control{
+      seq{
+          reset_bt_reg;
+          block_transfer;
+      }
+  }
+}
+
+
+
+//TODO(nathanielnrn): this is axi_wrapper, prefer to use @toplevel attribute but its not working
+// See individual channel components for explanations of signals
+component main(
```

Comment:
It occurs to me that, hopefully, this can be a big win for generator-ization: one needn't write out each signal in the top-level wrapper by hand; instead, the generator can literally do something like:

```
IN_SIGNALS = ['ARESET', 'ARREADY', 'RVALID', ...]
for mem in memories:
    for signal in IN_SIGNALS:
        yield f'{mem}_{signal}'
```